presto-main/src/main/java/io/prestosql/connector/system/RuleStatsSystemTable.java

```diff
@@ -0,0 +1,120 @@
+    private final ConnectorTableMetadata ruleStatsTable;
+    private final Optional<RuleStatsRecorder> ruleStatsRecorder;
+
+    @Inject
+    public RuleStatsSystemTable(Optional<RuleStatsRecorder> ruleStatsRecorder, Metadata metadata)
+    {
+        this.ruleStatsRecorder = requireNonNull(ruleStatsRecorder, "ruleStatsRecorder is null");
+        requireNonNull(metadata, "metadata is null");
+
+        this.ruleStatsTable = tableMetadataBuilder(TABLE_NAME)
+                .column("rule_name", VARCHAR)
+                .column("invocations", BIGINT)
+                .column("matches", BIGINT)
+                .column("failures", BIGINT)
+                .column("average_time", DOUBLE)
+                .column("time_distribution_percentiles", new InternalTypeManager(metadata).getType(mapType(DOUBLE.getTypeSignature(), DOUBLE.getTypeSignature())))
+                .column("time_unit", VARCHAR)
+                .build();
+    }
+
+    @Override
+    public Distribution getDistribution()
+    {
+        return SINGLE_COORDINATOR;
+    }
+
+    @Override
+    public ConnectorTableMetadata getTableMetadata()
+    {
+        return ruleStatsTable;
+    }
+
+    @Override
+    public ConnectorPageSource pageSource(ConnectorTransactionHandle transactionHandle, ConnectorSession session, TupleDomain<Integer> constraint)
+    {
+        checkState(ruleStatsRecorder.isPresent(), "Rule stats system table can return results only on coordinator");
+        Map<Class<?>, RuleStats> ruleStats = ruleStatsRecorder.get().getStats();
+
+        int positionCount = ruleStats.size();
+        Map<String, BlockBuilder> blockBuilders = ruleStatsTable.getColumns().stream()
+                .collect(toImmutableMap(ColumnMetadata::getName, column -> column.getType().createBlockBuilder(null, positionCount)));
+
+        for (Map.Entry<Class<?>, RuleStats> entry : ruleStats.entrySet()) {
+            RuleStats stats = entry.getValue();
+
+            VARCHAR.writeString(blockBuilders.get("rule_name"), entry.getKey().getSimpleName());
+            BIGINT.writeLong(blockBuilders.get("invocations"), stats.getInvocations());
+            BIGINT.writeLong(blockBuilders.get("matches"), stats.getHits());
+            BIGINT.writeLong(blockBuilders.get("failures"), stats.getFailures());
+            DOUBLE.writeDouble(blockBuilders.get("average_time"), stats.getTime().getAvg());
+
+            BlockBuilder mapWriter = blockBuilders.get("time_distribution_percentiles").beginBlockEntry();
+            for (Map.Entry<Double, Double> percentile : stats.getTime().getPercentiles().entrySet()) {
+                DOUBLE.writeDouble(mapWriter, percentile.getKey());
+                DOUBLE.writeDouble(mapWriter, percentile.getValue());
+            }
+            blockBuilders.get("time_distribution_percentiles").closeEntry();
+
+            VARCHAR.writeString(blockBuilders.get("time_unit"), stats.getTime().getUnit().name());
```

Comment:
Let's remove this column. All measurements are in the same unit, so we don't need to clarify what the unit is per row.